src/Lexer.ts

```diff
@@ -0,0 +1,98 @@
+// @author Anush kumar N <anush.kumar@codingmart.com>
+
+import TokenType from './Token';
+
+type arrayOfString = Array<string>;
+type literalWithType = { type: Symbol; literal: string | undefined };
+type identifyLiteralReturnType = { token: literalWithType; updatedPos: number | undefined };
+
+// Performs lexical analysis on raw string input.
+function* Lexer(input: string) {
+  const inputAsArr: arrayOfString = Array.from(input);
+  inputAsArr.push(TokenType.eof.toString());
+
+  let position: number = 0;
+  for (; position < inputAsArr.length; position++) {
+    const str: string = inputAsArr[position];
+    const { token, updatedPos } = identifyLiteral({ str, position, inputAsArr });
+    updatedPos && (position = updatedPos);
+
+    yield token;
+  }
+}
+
+// identifyLiteral helps to identify literals i.e either the token, number or identifier
+type identifyLiteralParams = {
+  str: string;
+  position: number;
+  inputAsArr: arrayOfString;
+};
+const identifyLiteral = (data: identifyLiteralParams): identifyLiteralReturnType => {
+  const { str, position, inputAsArr } = data;
+  let token, updatedPos;
+
+  switch (str) {
+    case '.':
+      token = { type: TokenType.dot, literal: str };
+      break;
+    case '[':
+      token = { type: TokenType.lbracket, literal: str };
+      break;
+    case ']':
+      token = { type: TokenType.rbracket, literal: str };
+      break;
+    default:
+      if (isNumeric(str)) {
+        const { updatedPosInt, int } = readNumber(position, inputAsArr) || {};
+        updatedPosInt && (updatedPos = updatedPosInt);
+        token = { type: TokenType.int, literal: int };
+      } else {
+        const { updatedPosIdent, ident } = readIdentifier(position, inputAsArr) || {};
+        updatedPosIdent && (updatedPos = updatedPosIdent);
+        token = { type: TokenType.ident, literal: ident };
+      }
```

Comment:
There is way too much cleverness happening here. Can you make it clearer and more explicit?

Reply:
sure i have updated the code